File: app/src/main/java/co/tiagoaguiar/codelab/myapplication/adapter/MainAdapter.java
package co.tiagoaguiar.codelab.myapplication.adapter;

import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.LinearLayout;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

import java.util.List;

import co.tiagoaguiar.codelab.myapplication.MainItem;
import co.tiagoaguiar.codelab.myapplication.OnItemClickListener;
import co.tiagoaguiar.codelab.myapplication.R;

public class MainAdapter extends RecyclerView.Adapter<MainAdapter.MainViewHolder> {
    private List<MainItem> mainItems;
    private OnItemClickListener listener;

    public MainAdapter(List<MainItem> mainItems) {
        this.mainItems = mainItems;
    }

    public void setListener(OnItemClickListener listener) {
        this.listener = listener;
    }

    @NonNull
    @Override
    public MainViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        return new MainViewHolder(
                LayoutInflater.from(parent.getContext())
                        .inflate(R.layout.main_item, parent, false)
        );
    }


    @Override
    public void onBindViewHolder(@NonNull MainViewHolder holder, int position) {
        MainItem mainItemCurrent = mainItems.get(position);
        holder.bind(mainItemCurrent);

    }

    @Override
    public int getItemCount() {
        return mainItems.size();
    }

    // Entenda como sendo a VIEW DA CELULA que está dentro da RecyclerView
    protected class MainViewHolder extends RecyclerView.ViewHolder {

        public MainViewHolder(@NonNull View itemView) {
            super(itemView);
        }

        public void bind(MainItem item) {
            TextView txtName = itemView.findViewById(R.id.item_txt_name);
            ImageView imgIcon = itemView.findViewById(R.id.item_img_icon);
            LinearLayout btnImc = (LinearLayout) itemView.findViewById(R.id.btn_imc);

            btnImc.setOnClickListener(v -> {
                listener.onClick(item.getId());
            });

            txtName.setText(item.getTextStringId());
            imgIcon.setImageResource(item.getDrawableId());
            btnImc.setBackgroundColor(item.getColor());
        }
    }
}

File: app/src/main/java/co/tiagoaguiar/fitnesstracker/adapter/MainAdapter.kt
package co.tiagoaguiar.fitnesstracker.adapter

import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.TextView
import androidx.cardview.widget.CardView
import androidx.recyclerview.widget.RecyclerView
import co.tiagoaguiar.fitnesstracker.R
import co.tiagoaguiar.fitnesstracker.adapter.item.MainItem

class MainAdapter(
    private val mainItems: List<MainItem>,
//    private val onItemClickListener: OnItemClickListener
    private val onItemClickListener: (Int)-> Unit
) : RecyclerView.Adapter<MainAdapter.MainViewHolder>() {
    // 1- Qual é o layout XML da célula específica(item)
    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): MainViewHolder {
        val view = LayoutInflater.from(parent.context)
            .inflate(R.layout.main_item, parent, false)
        return MainViewHolder(view)
    }

    // 2- Disparado toda vez que houver uma rolagem na tela e for necessário trocar o conteúdo
    //da célula
    override fun onBindViewHolder(holder: MainViewHolder, position: Int) {
        val itemCurrent = mainItems[position]
        holder.bind(itemCurrent)
    }

    // 3 - Quantas células essa listagem terá
    override fun getItemCount(): Int {
        return mainItems.size
    }

    inner class MainViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
        fun bind(item: MainItem) {
            val img: ImageView = itemView.findViewById(R.id.item_img_icon)
            val name: TextView = itemView.findViewById(R.id.item_txt_name)
            val card: CardView = itemView.findViewById(R.id.item_cardview)

            img.setImageResource(item.drawableId)
            name.setText(item.textStringId)

            card.setOnClickListener {
                //Aqui ele é uma interface
                // onItemClickListener.onClick(item.id)

                // aqui é uma função
                onItemClickListener.invoke(item.id)
            }

            // 3 maneiras de escutar eventos de click usando células (viewholder) activities
            // 1. impl interface
            // 2. objetos anonimos
            // 3. funcional
        }
    }

}